File: src/wuyu/protocol/jsonrpc.py
"""JSON-RPC message types for the Codex App Server protocol.

Note: The Codex App Server uses a non-standard JSON-RPC 2.0 — it omits the
"jsonrpc": "2.0" field. Messages are one of: Request, Notification, Response, Error.
"""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel

from wuyu.protocol.types import RequestId


class JSONRPCRequest(BaseModel):
    """A request from client to server (or server to client for approvals).
    Expects a response with the same id."""

    id: RequestId
    method: str
    params: dict[str, Any] | None = None


class JSONRPCNotification(BaseModel):
    """A fire-and-forget message (no id, no response expected)."""

    method: str
    params: dict[str, Any] | None = None


class JSONRPCError(BaseModel):
    """Error detail within an error response."""

    code: int
    message: str
    data: Any | None = None


class JSONRPCResponse(BaseModel):
    """Successful response to a request."""

    id: RequestId
    result: Any = None


class JSONRPCErrorResponse(BaseModel):
    """Error response to a request."""

    id: RequestId
    error: JSONRPCError


# Discriminated union of all message types.
# We use a tagged union approach for deserialization — see codec.py for the
# parse logic that inspects fields to determine the variant.
JSONRPCMessage = JSONRPCRequest | JSONRPCNotification | JSONRPCResponse | JSONRPCErrorResponse


# Well-known error codes
ERROR_SERVER_OVERLOADED = -32001
ERROR_PARSE_ERROR = -32700
ERROR_INVALID_REQUEST = -32600
ERROR_METHOD_NOT_FOUND = -32601

File: wuyu_dart/lib/src/protocol/jsonrpc.dart
/// JSON-RPC message types for the Codex App Server protocol.
///
/// The Codex App Server uses a non-standard JSON-RPC 2.0 variant — it omits
/// the "jsonrpc": "2.0" field. Messages are classified by field presence:
///   - id + method  → Request
///   - method only  → Notification
///   - id + result  → Response
///   - id + error   → ErrorResponse
library;

/// A request from client to server (or server to client for approvals).
/// Expects a response with the same id.
final class JsonRpcRequest {
  final Object id; // String or int
  final String method;
  final Map<String, Object?>? params;

  const JsonRpcRequest({required this.id, required this.method, this.params});
}

/// A fire-and-forget message — no id, no response expected.
final class JsonRpcNotification {
  final String method;
  final Map<String, Object?>? params;

  const JsonRpcNotification({required this.method, this.params});
}

/// Successful response to a request.
final class JsonRpcResponse {
  final Object id; // String or int
  final Object? result;

  const JsonRpcResponse({required this.id, this.result});
}

/// Error detail within an error response.
final class JsonRpcError {
  final int code;
  final String message;
  final Object? data;

  const JsonRpcError({required this.code, required this.message, this.data});
}

/// Error response to a request.
final class JsonRpcErrorResponse {
  final Object id; // String or int
  final JsonRpcError error;

  const JsonRpcErrorResponse({required this.id, required this.error});
}

// Well-known error codes
const errorServerOverloaded = -32001;
const errorParseError = -32700;
const errorInvalidRequest = -32600;
const errorMethodNotFound = -32601;
